[Feat][Core/Dashboard] Add SubprocessModules to the Dashboard routes, and convert HealthzHead

MortalHappiness opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
Based on https://github.com/ray-project/ray/pull/51172 and https://github.com/ray-project/ray/pull/49864 to convert `HealthzHead` to subprocess dashboard module.

- Integrate SubprocessModule into Dashboard. Now it can load SubprocessModules.
- The support is only in non-minimal. For a minimal Ray, no SubprocessModule are loaded, but others minimal `DashboardHeadModule`s can still work.
- Converted HealthzHead to non-minimal. It does not do anything in minimal anyway.
- Optimized usage_stats.py to use multi_get to save number of rpcs.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jjyao commented on python/ray/_private/ray_constants.py:
```diff
@@ -539,3 +539,10 @@ def gcs_actor_scheduling_enabled():
+# Skip invoking `wait_for_module_ready()` for `SubprocessModuleHandle` in the dashboard
+# when loading modules. Only used for testing because calling it with CliRunner
+# in the test environment will hang.
+RAY_SKIP_INVOKE_SUBPROCESS_DASHBOARD_WAIT_FOR_MODULE_READY = os.environ.get(
```
Let's figure out why.

MortalHappiness replied:
Thanks. Let me provide you a reproducible script:

```python
def test_dashboard(tmpdir):
    head = DashboardHead(
        http_host="127.0.0.1",
        http_port=8265,
        http_port_retries=1,
        node_ip_address="127.0.0.1",
        gcs_address="127.0.0.1:6379",
        cluster_id_hex=ray.ClusterID.from_random().hex(),
        grpc_port=0,
        log_dir=str(tmpdir),
        logging_level=ray_constants.LOGGER_LEVEL,
        logging_format=ray_constants.LOGGER_FORMAT,
        logging_filename=dashboard_consts.DASHBOARD_LOG_FILENAME,
        logging_rotate_bytes=ray_constants.LOGGING_ROTATE_BYTES,
        logging_rotate_backup_count=ray_constants.LOGGING_ROTATE_BACKUP_COUNT,
        temp_dir=str(tmpdir),
        session_dir=str(tmpdir),
        minimal=False,
        serve_frontend=True,
    )
    head._load_modules()
```

Running this test, `module.py` will fail in the line `await site.start()`. 

Error messages look like this:

```
2025-03-14 10:24:45,200	ERROR module.py:184 -- Error creating module HealthzHead
Traceback (most recent call last):
  File "/home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py", line 180, in run_module_inner
    await module.start_server()
  File "/home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py", line 126, in start_server
    await site.start()
  File "/home/mortalhappiness/miniforge3/envs/ray/lib/python3.12/site-packages/aiohttp/web_runner.py", line 162, in start
    self._server = await loop.create_unix_server(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mortalhappiness/miniforge3/envs/ray/lib/python3.12/asyncio/unix_events.py", line 319, in create_unix_server
    sock.bind(path)
FileNotFoundError: [Errno 2] No such file or directory
2025-03-14 10:24:45,201	ERROR base_events.py:1820 -- Task exception was never retrieved
future: <Task finished name='Task-1' coro=<run_module_inner() done, defined at /home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py:159> exception=FileNotFoundError(2, 'No such file or directory')>
Traceback (most recent call last):
  File "/home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py", line 185, in run_module_inner
    raise e
  File "/home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py", line 180, in run_module_inner
    await module.start_server()
  File "/home/mortalhappiness/Projects/opensource/ray-project/ray/python/ray/dashboard/subprocesses/module.py", line 126, in start_server
    await site.start()
  File "/home/mortalhappiness/miniforge3/envs/ray/lib/python3.12/site-packages/aiohttp/web_runner.py", line 162, in start
    self._server = await loop.create_unix_server(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mortalhappiness/miniforge3/envs/ray/lib/python3.12/asyncio/unix_events.py", line 319, in create_unix_server
    sock.bind(path)
FileNotFoundError: [Errno 2] No such file or directory
```

MortalHappiness replied:
Fixed

jjyao commented on python/ray/_private/usage/usage_lib.py:
```diff
@@ -561,8 +561,8 @@ def get_total_num_nodes_to_report(gcs_client, timeout=None) -> Optional[int]:
                 total_num_nodes += 1
         return total_num_nodes
-    except Exception as e:
-        logger.info(f"Faile to query number of nodes in the cluster: {e}")
```
It's done this way intentionally because we want it to be an INFO level log.

jjyao commented on python/ray/dashboard/dashboard.py:
```diff
@@ -51,7 +51,12 @@ def __init__(
         grpc_port: int,
         node_ip_address: str,
-        log_dir: str = None,
+        log_dir: str,
```
Update the above docstring

MortalHappiness commented on python/ray/dashboard/subprocesses/utils.py:
```diff
@@ -12,17 +12,38 @@ class ResponseType(enum.Enum):
+    return f"{stem}_{module_name}_{incarnation}{extension}"
+
+
+def get_socket_path(socket_dir: str, module_name: str) -> str:
+    socket_path = os.path.join(socket_dir, "dashboard_" + module_name)
+    # The max length of a Unix socket path is 108 bytes.
+    if len(socket_path) > 108:
+        # Discard the "dashboard_" prefix.
+        socket_path = os.path.join(socket_dir, module_name)
+    if len(socket_path) > 108:
+        raise ValueError(
+            f"Socket path {socket_path} is too long. "
+            "Please use a shorter module name or a shorter socket_dir."
+        )
+    return socket_path
```
In tests, the `socket_dir` is very long because it is the `tmp_path` pytest fixture plus the session folder name, which contains a timestamp. It looks like this:

`/tmp/4163fbc40ed74552a12e343de9b35fc9/session_2025-03-14_11-09-51_033910_1341615/sockets/`

The maximum length for a Unix socket is 108 bytes. Therefore, if the total `socket_path` exceeds 108 characters, we discard the `dashboard_` prefix.

jjyao commented on python/ray/dashboard/head.py:
```diff
@@ -198,27 +253,78 @@ def _load_modules(self, modules_to_load: Optional[Set[str]] = None):
+                cls for cls in subprocess_cls_list if cls.__name__ in modules_to_load
+            ]
+
+        # Parallel start all modules.
+        for cls in subprocess_cls_list:
+            logger.info(f"Loading {SubprocessModule.__name__}: {cls}.")
+            handle = SubprocessModuleHandle(loop, cls, config)
+            handle.start_module()
+            handles.append(handle)
+
+        # Wait for all modules to be ready.
+        for handle in handles:
+            handle.wait_for_module_ready()
```
I think this should be out of `_load_subprocess_module_handles`

jjyao commented on python/ray/dashboard/modules/healthz/healthz_head.py:
```diff
@@ -33,9 +35,5 @@ async def health_check(self, req: Request) -> Response:
         return HTTPServiceUnavailable(reason="Health check failed")
 
-    async def run(self, server):
+    async def init(self):
         pass
```
Will we have a module that implements a non-empty `init()`?

MortalHappiness replied:
Yes. For example, `EventHead` https://github.com/ray-project/ray/pull/50428/files#diff-4da7e8e823b529b1cf58a5b1ee8acfac42c6390789f084730589c974a9fb1267L203-R221

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -201,12 +205,13 @@ async def _do_periodic_health_check(self):
             except Exception:
                 filename = module_logging_filename(
-                    self.module_cls.__name__, self.config.logging_filename
+                    self.module_cls.__name__, incarnation, self.config.logging_filename
```
Let's not have the `incarnation` as part of the log name. Can always just be dashboard_xxxx.log

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -47,12 +52,17 @@ class SubprocessModule(abc.ABC):
+        self._parent_process_pid = parent_process_pid
+        # Lazy init
+        self._gcs_aio_client = None
+        self._parent_process_death_detection_task = None
```
No need to have the parent death detection since the subprocess is started with `daemon=True`.

jjyao commented on python/ray/dashboard/subprocesses/utils.py:
```diff
@@ -12,17 +12,38 @@ class ResponseType(enum.Enum):
+
+def get_socket_path(socket_dir: str, module_name: str) -> str:
+    socket_path = os.path.join(socket_dir, "dashboard_" + module_name)
+    # The max length of a Unix socket path is 108 bytes.
+    if len(socket_path) > 108:
+        # Discard the "dashboard_" prefix.
+        socket_path = os.path.join(socket_dir, module_name)
+    if len(socket_path) > 108:
+        raise ValueError(
+            f"Socket path {socket_path} is too long. "
+            "Please use a shorter module name or a shorter socket_dir."
+        )
```
Share the same logic as `_prepare_socket_file`

jjyao commented on python/ray/train/BUILD:
```diff
@@ -749,8 +749,14 @@ py_test(
     srcs = ["tests/test_state_export.py"],
-    tags = ["team:ml", "exclusive"],
-    deps = [":train_lib", ":conftest"]
+    tags = [
```
Let's not change this in this file

MortalHappiness replied:
Ah, it is complained by the linter in CI.

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -106,35 +116,63 @@ async def start_server(self):
             site = aiohttp.web.NamedPipeSite(runner, socket_path)
         else:
             site = aiohttp.web.UnixSite(runner, socket_path)
+        logger.info(f"ABC: {socket_path=}, {len(socket_path)}")
```
?

jjyao commented on python/ray/dashboard/subprocesses/utils.py:
```diff
@@ -12,17 +12,38 @@ class ResponseType(enum.Enum):
+
+
+def get_socket_path(socket_dir: str, module_name: str) -> str:
+    socket_path = os.path.join(socket_dir, "dashboard_" + module_name)
```
```suggestion
    socket_path = os.path.join(socket_dir, "dash_" + module_name)
```

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1001,12 +1005,7 @@ def _prepare_socket_file(self, socket_path: str, default_prefix: str):
-                raise OSError(
-                    f"AF_UNIX path length cannot exceed {maxlen} bytes: {result!r}"
-                )
+            validate_socket_filename(result.split("://", 1)[-1])
```
```suggestion
            validate_socket_filepath(result.split("://", 1)[-1])
```

jjyao commented on python/ray/_private/utils.py:
```diff
@@ -2078,3 +2078,18 @@ def get_current_node_cpu_model_name() -> Optional[str]:
         return None
+
+
+def validate_socket_filename(filename: str):
```
```suggestion
def validate_socket_filename(filepath: str):
```

jjyao commented on python/ray/dashboard/head.py:
```diff
@@ -328,6 +432,13 @@ async def _async_notify():
             namespace=ray_constants.KV_NAMESPACE_DASHBOARD,
         )
 
+        # Parallel start all subprocess modules.
+        for handle in subprocess_module_handles:
+            handle.start_module()
+        # Wait for all subprocess modules to be ready.
+        for handle in subprocess_module_handles:
+            handle.wait_for_module_ready()
+
```
We should do this before we start the server like before `_configure_http_server`?

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -116,14 +117,19 @@ def start_module(self):
                 self.module_cls,
                 self.config,
+                self.incarnation,
+                os.getpid(),
```
no need to pass in the parent pid anymore?

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -148,12 +155,21 @@ async def run_module_inner(
     """
     Entrypoint for a subprocess module.
+
+    parent_process_pid: Used to detect if the parent process died every 1s. If it does,
+    the module will exit.
```
no longer needed

MortalHappiness commented:
Added back `_parent_process_death_detection_task` and removed the `daemon` attribute from the subprocess because I found it hangs when calling `run_string_as_driver` in tests.

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1001,12 +1005,7 @@ def _prepare_socket_file(self, socket_path: str, default_prefix: str):
-                raise OSError(
-                    f"AF_UNIX path length cannot exceed {maxlen} bytes: {result!r}"
-                )
+            validate_socket_filepath(result.split("://", 1)[-1])
```
Do we need to keep `.encode("utf-8")`?

MortalHappiness replied:
Do we want to use non-ASCII chracters for filename? If we only use ASCII characters for filename, then the length of string and the length of the encoded bytes will be the same.

jjyao replied:
I don't know the history, I think it's safer to keep it.

MortalHappiness replied:
Added back `.encode` inside `validate_socket_filepath`.

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -47,12 +52,30 @@ class SubprocessModule(abc.ABC):
+        Detect parent process death by checking if ppid is still the same.
+        """
+        while True:
+            ppid = os.getppid()
```
```
Return the parent’s process id. When the parent process has exited, on Unix the id returned is the one of the init process (1), on Windows it is still the same id, which may be already reused by another process.
```

This approach won't work on Windows.

Try `multiprocessing.parent_process().is_alive()` or pipe.

MortalHappiness replied:
Use `psutil` instead.

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -47,12 +52,38 @@ class SubprocessModule(abc.ABC):
         :param dashboard_head: The DashboardHead instance.
         """
         self._config = config
+        self._parent_process = psutil.Process(parent_process_pid)
```
What if parent is already dead and the pid is reused?

MortalHappiness replied:
This is in the constructor of the child process. So do you mean that the parent process dies immediately even before the child process constructor has run?

MortalHappiness replied:
I've changed the implementation to use `multiprocessing.parent_process().is_alive()` because I found that the `parent_process()` call returns a variable maintained by Python, so it should be able to identify the parent process calling `multiprocessing`.  

https://github.com/python/cpython/blob/a936af924efc6e2fb59e27990dcd905b7819470a/Lib/multiprocessing/process.py#L51-L55

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -53,6 +57,22 @@ def __init__(
+
+    async def _detect_parent_process_death(self):
+        """
+        Detect parent process death by checking if ppid is still the same.
```
Needs to update the comment

MortalHappiness replied:
updated
